### ISSUE TITLE
TCP client howto missing reactor.run() in one code sample

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3834]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#3834|
|Type|enhancement|
|Created|2009-05-13 11:39:48Z|
|Branch|https://github.com/twisted/twisted/tree/examples-reactor-run-3834-2|

In the "Simple, single-use clients" section there's a full example, but it's missing a `reactor.run()`.

<details><summary>Searchable metadata</summary>

```
trac-id__3834 3834
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_examples_reactor_run_3834_2 branches/examples-reactor-run-3834-2
branch_author__thijs thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation__easy documentation, easy
time__1242214788000000 1242214788000000
changetime__1297710357000000 1297710357000000
version__None None
owner__thijs thijs
cc__thijs cc__mwh cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

http://twistedmatrix.com/projects/core/documentation/howto/rdbms.html#auto5 is missing it too.  I bet lots of other docs are missing it, too.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @thijstriemstra|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

(In [27112]) Branching to 'examples-reactor-run-3834'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra removed owner|
|-|-|

Replying to [exarkun](#note_2):
> http://twistedmatrix.com/projects/core/documentation/howto/rdbms.html#auto5 is missing it too.  I bet lots of other docs are missing it, too.

Seems this is missing more than just as a reactor.run() call but also the imports, which is true for all of the examples on that page. If you can fix these imports, please do. Putting up the changes for the original ticket description for review.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_6"></a>|@mwhudson set owner to @thijstriemstra|
|-|-|

I'm not sure adding reactor.run() to the example in clients.xhtml without adding a reactor.stop() as well.  It's also a slightly odd example.  Do you think changing it to this:

```
def gotProtocol(p):
    p.sendMessage("Hello")
    defer.maybeDeferred(p.transport.loseConnection).addCallback(reactor.stop)
```

would make more sense?

All the other changes look good.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_7"></a>|@mwhudson commented|
|-|-|

Oops, I forgot something: thanks for working on this!

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra removed owner|
|-|-|

Replying to [mwh](#note_6):
> I'm not sure adding reactor.run() to the example in clients.xhtml without adding a reactor.stop() as well.  It's also a slightly odd example.  Do you think changing it to this:
>
> def gotProtocol(p):
>     p.sendMessage("Hello")
>     defer.maybeDeferred(p.transport.loseConnection).addCallback(reactor.stop)
>
> would make more sense?

Well, if I do that and run the example it throws this error:

```
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.internet.error.ConnectionRefusedError: Connection was refused by other side: 61: Connection refused.
```

Any idea?

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_9"></a>|@mwhudson commented|
|-|-|

Well, the existing code does that if nothing is listening.

I'm having a hard time getting this correct, so I don't know if I'm a good person to review this branch :)  I've got as far as this:

```
from twisted.internet import reactor
from twisted.internet.protocol import Protocol, ClientCreator
from twisted.python import log

class Greeter(Protocol):
    def sendMessage(self, msg):
        self.transport.write("MESSAGE %s\n" % msg)

def gotProtocol(p):
    p.sendMessage("Hello")
    return p.transport.loseConnection()

c = ClientCreator(reactor, Greeter)
d = c.connectTCP("localhost", 1234)
d.addCallback(gotProtocol).addErrback(log.err).addCallback(
    lambda ignored: reactor.stop())
reactor.run()
```

But the reactor.stop() gets called before the data is sent.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_10"></a>|@mwhudson set owner to @thijstriemstra|
|-|-|

I should have moved this out of the review state.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

assigning to exarkun for help/advice with this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Since the code is already doing arbitrary stuff with delayed calls, I think you just want to call `reactor.stop` after some arbitrary delay that's long enough to let the message be sent most of the time.

That, or make it actually work right, and put the reactor.stop call in the protocol's connection lost (which defeats the purpose of the example, which seems to be about using protocol instances from code that's not part of the protocol class).

So yea, `reactor.callLater(3, reactor.stop)`.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra commented|
|-|-|

(In [30553]) correct stop call, add news file. refs [#3834](https://github.com/twisted/twisted/issues/3834)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [exarkun](#note_12):
> So yea, `reactor.callLater(3, reactor.stop)`.

Merged forward and replaced that call.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

There's a conflict in `clients.xhtml` now.  Can you merge forward and resolve it?  Also, the news file doesn't reflect the style described on ReviewProcess#Newsfiles.  Can you re-arrange it so it does, and mention the udp and client howtos specifically?

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Thanks for the review, i'll look into it.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra commented|
|-|-|

(In [30594]) Branching to 'examples-reactor-run-3834-2'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_19"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Applied the changes in a new branch. Doc builder is [happy](http://buildbot.twistedmatrix.com/builders/documentation/builds/902).

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @jerub<br>@jerub set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_21"></a>|@jerub set owner to @thijstriemstra<br>@jerub set status to `new`|
|-|-|

I've applied this branch to my working copy, read through the examples in the two affected files and while 2 of the examples are runnable, the third simply has an ellipsis and is not runnable - I cut+paste what was required from the other example to splice together something that would work, and it did.

What I'm saying is, I'm glad these examples have been improved, I've tested them, and they work, please merge.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_23"></a>|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

Thanks for the review.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_24"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [30759]) Added missing reactor.run() calls in the UDP and client howto documents.

Author: thijs
Reviewer: mwh, exarkun, jerub
Fixes: [#3834](https://github.com/twisted/twisted/issues/3834)
